File: platforms/ios/Packages/FunputKit/Sources/PersonalSuggestions/Service/PersonalSuggestionService.swift
import Foundation
import KeyboardInput
import KeyboardLayout
import KeyboardRenderer

@MainActor
public final class PersonalSuggestionService {
    public var onCandidates: ((UInt64, [KeyboardSuggestionCandidate]) -> Void)?
    private let workerFactory: PersonalSuggestionWorkerFactory
    private var worker: (any PersonalSuggestionWorking)?
    private var workerConfiguration = PersonalSuggestionWorkerConfiguration(
        enabled: false,
        hasFullAccess: false,
        resetToken: nil
    )
    private(set) var generation: UInt64 = 0
    private var activationGeneration: UInt64 = 0
    private(set) var prefix = ""
    private var context: String?
    /// Whether the bar is currently answering a context rather than a prefix.
    private var predicting = false
    /// As the engine returned them. `visibleCandidates` puts them in the user's case.
    var received: [PersonalSuggestionCandidate] = []
    var shift: ShiftState = .lowercase
    /// Last drawn; `nil` after a reconfigure, so reactivation always redraws.
    private var published: [String]?

    public init(workerFactory: @escaping PersonalSuggestionWorkerFactory) {
        self.workerFactory = workerFactory
    }

    public func configure(
        enabled: Bool,
        hasFullAccess: Bool,
        resetToken: UUID?,
        activationGeneration: UInt64
    ) {
        workerConfiguration = .init(
            enabled: enabled,
            hasFullAccess: hasFullAccess,
            resetToken: resetToken
        )
        self.activationGeneration = activationGeneration
        worker?.configure(workerConfiguration)
        published = nil
        invalidateAndPublish()
    }

    public func update(
        _ update: KeyboardSuggestionInputUpdate,
        canQuery: Bool,
        shift: ShiftState = .lowercase
    ) {
        let signpostID = PersonalSuggestionServiceSignposts.beginDispatch()
        defer {
            PersonalSuggestionServiceSignposts.endDispatch(
                signpostID,
                generation: generation
            )
        }
        if let token = update.completedToken, workerConfiguration.enabled {
            ensureWorker()?.learn(token, after: context)
        }
        context = update.context
        self.shift = shift
        let allowed = canQuery && workerConfiguration.enabled
        let next = allowed && update.prefix.count >= 2 ? update.prefix : ""
        // A prediction answers a context with no prefix at all. One character is
        // neither a prefix worth completing nor a word boundary, and stays out.
        let predicts = allowed && update.prefix.isEmpty && context != nil
        if update.completedToken == nil, next == prefix, predicts == predicting { return }
        generation &+= 1
        prefix = next
        predicting = predicts
        received = []
        publish([])
        guard !prefix.isEmpty || predicting else { return }
        ensureWorker()?.query(.init(prefix: prefix, generation: generation, context: context))
    }

    public func acceptance(
        for candidate: KeyboardSuggestionCandidate
    ) -> (String, String)? {
        // An empty prefix is the ordinary shape of accepting a prediction: it
        // replaces nothing and inserts a word.
        guard candidate.generation == generation,
              visibleCandidates.contains(candidate)
        else { return nil }
        return (prefix, candidate.text)
    }

    public func clear() { invalidateAndPublish() }
    public func flush() { worker?.flush() }

    private func ensureWorker() -> (any PersonalSuggestionWorking)? {
        guard workerConfiguration.enabled else { return nil }
        if let worker { return worker }
        let created = workerFactory { [weak self] request, candidates in
            Task { @MainActor [weak self] in
                self?.receive(request: request, candidates: candidates)
            }
        }
        PersonalSuggestionServiceSignposts.recordWorkerCreated()
        created.configure(workerConfiguration)
        worker = created
        return created
    }

    private func invalidateAndPublish() {
        let needsPublish = !prefix.isEmpty || predicting || !received.isEmpty
        generation &+= 1
        prefix = ""
        context = nil
        predicting = false
        received = []
        if needsPublish { publish([]) }
    }

    private func receive(
        request: PersonalSuggestionQueryRequest,
        candidates: [PersonalSuggestionCandidate]
    ) {
        guard workerConfiguration.enabled,
              request.generation == generation, request.prefix == prefix
        else { return }
        received = candidates
        let values = visibleCandidates
        publish(values)
        PersonalSuggestionServiceSignposts.recordUI(
            generation: generation,
            count: values.count
        )
    }

    func publish(_ candidates: [KeyboardSuggestionCandidate]) {
        // Prediction adds a bar update after every space, most of them one empty
        // list after another. Comparing here keeps that cost off UIKit.
        let texts = candidates.map(\.text)
        guard texts != published else { return }
        published = texts
        onCandidates?(activationGeneration, candidates)
    }
}

File: platforms/android/ime/src/main/java/app/funput/funput/ime/suggestions/service/PersonalSuggestionService.kt
package app.funput.funput.ime.suggestions

import android.content.Context
import app.funput.funput.ime.editing.AuthoredSuggestionUpdate
import app.funput.funput.ime.editing.EditorInfoPolicy
import app.funput.funput.ime.editing.ImeKeyActionHandler
import app.funput.funput.ime.editing.ImeSuggestionSource
import app.funput.funput.ime.settings.PersonalSuggestionPreferences
import app.funput.funput.ime.settings.pendingReset
import app.funput.funput.keyboard.model.ShiftState
import app.funput.funput.keyboard.model.SuggestionSelection
import app.funput.funput.keyboard.ui.KeyboardPanel

internal class PersonalSuggestionService(
    context: Context,
    private val show: (List<String>) -> Unit,
    private val shift: () -> ShiftState = { ShiftState.OFF },
    private val acknowledgeReset: (String) -> Unit,
    createWorker: (Context, (PersonalSuggestionRequest, List<String>) -> Unit) -> PersonalSuggestionWorker = ::suggestionWorker,
) {
    private val worker = runCatching { createWorker(context, ::publish) }.getOrNull()
    private var preferences = PersonalSuggestionPreferences.Default
    private var policy = EditorInfoPolicy.Default
    private var panel = KeyboardPanel.LETTERS
    private var session = 0L
    private var generation = 0L
    private var prefix = ""
    /** The word the next one follows, when the tracker can vouch for one. */
    private var previousWord: String? = null
    /** Whether the bar is currently answering a context rather than a prefix. */
    private var predicting = false
    /** As the engine returned them; [candidates] is the same list in the user's case. */
    private var received = emptyList<String>()
    private var candidates = emptyList<String>()
    private var pendingReset: String? = null

    fun configure(value: PersonalSuggestionPreferences) {
        preferences = value
        val reset = value.pendingReset(pendingReset)
        if (reset != null && worker != null) {
            pendingReset = reset
            worker.reset {
                pendingReset = null
                clear()
                acknowledgeReset(reset)
            }
        }
        if (!value.enabled) clear()
    }

    fun start(value: EditorInfoPolicy) {
        session += 1
        policy = value
        panel = KeyboardPanel.LETTERS
        clear()
    }

    fun finish() {
        session += 1
        clear()
        worker?.flush()
    }

    fun updatePanel(value: KeyboardPanel) {
        panel = value
        if (value != KeyboardPanel.LETTERS) clear()
    }

    fun consume(update: AuthoredSuggestionUpdate) {
        update.completedToken
            ?.takeIf { eligible() && policy.allowsPersonalizedLearning }
            ?.let { worker?.learn(it, previousWord) }
        previousWord = update.context
        // A prediction answers a context with no prefix at all. One character is
        // neither a prefix worth completing nor a word boundary, and stays out.
        val predicts = eligible() && update.prefix.isEmpty() && previousWord != null
        if (update.completedToken == null && update.prefix == prefix && predicts == predicting) {
            // Shift arrives here: same words, new shape, nothing to ask the engine.
            return draw()
        }
        prefix = update.prefix
        predicting = predicts
        if (!eligible() || (!predicts && prefix.codePointCount(0, prefix.length) < MinimumPrefix)) {
            return clear()
        }
        generation += 1
        clearCandidates()
        worker?.query(PersonalSuggestionRequest(prefix, generation, session, previousWord))
    }

    fun select(selection: SuggestionSelection, handler: ImeKeyActionHandler): Boolean {
        val candidate = candidates.getOrNull(selection.index) ?: return reject()
        // An empty prefix is the ordinary shape of accepting a prediction: it
        // replaces nothing and inserts a word.
        if (candidate != selection.text || !eligible()) return reject()
        if (!handler.acceptSuggestion(candidate, prefix)) return reject()
        consume(handler.takeSuggestionUpdate())
        return true
    }

    fun flush() = worker?.flush()
    fun close() = worker?.close()

    private fun publish(request: PersonalSuggestionRequest, values: List<String>) {
        if (request.generation != generation || request.session != session || request.prefix != prefix) return
        received = values
        draw()
        suggestionCounter("SuggestionQueryToToolbarUs", (System.nanoTime() - request.startedNanos) / 1_000)
    }

    /** Cases what the engine gave us and shows it, keeping both lists in step: the
     *  accept path matches the tapped text against [candidates], so a bar drawn from a
     *  list nobody stored is a bar whose every tap is silently refused. */
    private fun draw() {
        if (!eligible()) return
        val next = received.map { PersonalSuggestionCasing.apply(it, prefix, shift()) }
        if (next == candidates) return
        candidates = next
        runCatching { show(candidates) }
    }

    private fun eligible() = preferences.enabled && policy.allowsPersonalSuggestions &&
        policy.suggestionSource == ImeSuggestionSource.FUNPUT &&
        panel == KeyboardPanel.LETTERS

    private fun reject(): Boolean = false.also { clear() }

    private fun clear() {
        generation += 1
        prefix = ""
        previousWord = null
        predicting = false
        worker?.clearQueries()
        clearCandidates()
    }

    private fun clearCandidates() {
        // The raw list goes with them, or a later Shift would redraw what was dropped.
        received = emptyList()
        if (candidates.isEmpty()) return
        candidates = emptyList()
        runCatching { show(emptyList()) }
    }

    private companion object {
        const val MinimumPrefix = 2
    }
}
